java/src/main/native/src/multi_host_buffer_source.cpp

```diff
@@ -0,0 +1,165 @@
+    auto copy_size                 = std::min(buffer_left, bytes_left);
+    std::memcpy(dst, src, copy_size);
+    offset += copy_size;
+    dst += copy_size;
+    bytes_left -= copy_size;
+    ++buffer_index;
+  }
+  return size;
+}
+
+std::unique_ptr<cudf::io::datasource::buffer> multi_host_buffer_source::device_read(
+  size_t offset, size_t size, rmm::cuda_stream_view stream)
+{
+  if (size == 0) { return 0; }
+  if (offset < 0 || offset >= offsets_.back()) { throw std::runtime_error("bad offset"); }
+  if (offset + size > offsets_.back()) { throw std::runtime_error("read past end of file"); }
+  rmm::device_buffer buf(size, stream);
+  auto dst          = static_cast<uint8_t*>(buf.data());
+  auto buffer_index = locate_offset_index(offset);
+  auto bytes_left   = size;
+  while (bytes_left > 0) {
+    auto next_offset = offsets_[buffer_index + 1];
+    auto buffer_size = next_offset = offsets_[buffer_index];
+    auto buffer_left               = next_offset - offset;
+    auto src                       = addrs_[buffer_index] + buffer_size - buffer_left;
+    auto copy_size                 = std::min(buffer_left, bytes_left);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(dst, src, copy_size, cudaMemcpyHostToDevice, stream.value()));
+    offset += copy_size;
+    dst += copy_size;
+    bytes_left -= copy_size;
+    ++buffer_index;
+  }
+  return std::make_unique<owning_buffer<rmm::device_buffer>>(std::move(buf));
+}
+
+size_t multi_host_buffer_source::device_read(size_t offset,
+                                             size_t size,
+                                             uint8_t* dst,
+                                             rmm::cuda_stream_view stream)
+{
+  if (size == 0) { return 0; }
+  if (offset < 0 || offset >= offsets_.back()) { throw std::runtime_error("bad offset"); }
+  if (offset + size > offsets_.back()) { throw std::runtime_error("read past end of file"); }
+  auto buffer_index = locate_offset_index(offset);
+  auto bytes_left   = size;
+  while (bytes_left > 0) {
+    auto next_offset   = offsets_[buffer_index + 1];
+    auto buffer_left   = next_offset - offset;
+    auto buffer_offset = offset - offsets_[buffer_index];
+    auto src           = addrs_[buffer_index] + buffer_offset;
+    auto copy_size     = std::min(buffer_left, bytes_left);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(dst, src, copy_size, cudaMemcpyHostToDevice, stream.value()));
+    offset += copy_size;
+    dst += copy_size;
+    bytes_left -= copy_size;
+    ++buffer_index;
+  }
+  return size;
+}
+
+std::future<size_t> multi_host_buffer_source::device_read_async(size_t offset,
+                                                                size_t size,
+                                                                uint8_t* dst,
+                                                                rmm::cuda_stream_view stream)
+{
+  std::promise<size_t> p;
+  p.set_value(device_read(offset, size, dst, stream));
+  return p.get_future();
```

Comment:
@vuule is it safe to return from the future with device activity still potentially pending on the specified stream?  It wasn't clear to me from the documentation whether the caller of this method expects the device operation to be synchronously completed when this future completes.

Reply:
yeah, this is not really specified. I think letting the device side of the operation be performed after the future would not break anything, but it wasn't intended originally. How does this impact you here?

Reply:
I'd like to avoid unnecessary stream synchronization if cuio will guarantee that it will do any required stream synchronization after the future completes.  If that's not guaranteed then I'll need to add explicit stream synchronization per read operation for safety.

Reply:
The only thing we have to be careful about is keeping the host buffers around during the parquet read. I don't think we have a worry about that given the API.